platforms/android/tangram/src/main/java/com/mapzen/tangram/DefaultHttpHandler.java

```diff
@@ -0,0 +1,231 @@
+                .followRedirects(true)
+                .followSslRedirects(true)
+                .connectTimeout(10, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(30, TimeUnit.SECONDS);
+
+        if (directory != null && maxSize > 0) {
+            builder.cache(new Cache(directory, maxSize));
+        }
+
+        // Use specified policy or construct default if null.
+        cachePolicy = policy;
+        tileCacheControl = cacheControl;
+        if (cachePolicy == null || tileCacheControl == null) {
+            cachePolicy = new CachePolicy() {
+                @Override
+                public boolean apply(@NonNull final String url) {
+                    return false;
+                }
+            };
+        }
+
+        if (Build.VERSION.SDK_INT < 22) {
+            try {
+                final SSLContext sc = SSLContext.getInstance("TLSv1.2");
+                sc.init(null, null, null);
+                builder.sslSocketFactory(new Tls12SocketFactory(sc.getSocketFactory()));
+
+                final ConnectionSpec cs = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+                        .tlsVersions(TlsVersion.TLS_1_2)
+                        .build();
+
+                final List<ConnectionSpec> specs = new ArrayList<>();
+                specs.add(cs);
+                specs.add(ConnectionSpec.COMPATIBLE_TLS);
+                specs.add(ConnectionSpec.CLEARTEXT);
+
+                builder.connectionSpecs(specs);
+            } catch (final Exception exc) {
+                android.util.Log.e("Tangram", "Error while setting TLS 1.2", exc);
+            }
+        }
+
+        okClient = builder.build();
+    }
+
+    @Override
+    public Object startRequest(@NonNull final String url, @NonNull final HttpHandler.Callback cb) {
+        final HttpUrl httpUrl = HttpUrl.parse(url);
+        Call call = null;
+        if (httpUrl == null) {
+            cb.onFailure(new IOException("HttpUrl failed to parse url=" + url));
+        }
+        else {
+            // Construct okhttp3.Callback which forwards apt response calls to internal HttpResponse.Callback
+            final okhttp3.Callback callback = new okhttp3.Callback() {
+                @Override
+                public void onFailure(final Call call, final IOException e) {
+                    if (call.isCanceled()) {
+                        cb.onCancel();
+                        return;
+                    }
+                    cb.onFailure(e);
+                }
+
+                @Override
+                public void onResponse(final Call call, final Response response) throws IOException {
+                    try {
+                        final ResponseBody body = response.body();
+                        // TODO headers? cacheControl?
+                        cb.onResponse(response.code(), body.bytes(), response.headers().toMultimap());
```

Comment:
The control flow here should ensure that onResponse or onFailure is called. Even if an exception is thrown while reading the response we want to invoke the HttpHandler.Callback so that our native request is properly finalized.

Reply:
Actually this gets very interesting!
> Code that is writing the request body or reading the response body will suffer an IOException when its call is canceled

(https://github.com/square/okhttp/wiki/Calls#calls)

So maybe the `catch` path should call `onFailure`? Or should check the cancelled status of the call and call `onCancel`?

Reply:
Yeah probably makes sense to check for cancellation in the catch block and aptly call `onCancel` or `onFailure` otherwise.